kubeflow/trainer/rhai/transformers.py

```
@@ -475,28 +518,121 @@ def relative_update(self, inc=1):
+                            recursive=True,
+                            callback=self._cloud_storage_progress_callback("Upload", local_size),
+                        )
+
+                        print(f"[Kubeflow] Upload complete: {checkpoint_name}", flush=True)
+
+                    except Exception as e:
+                        raise RuntimeError(
+                            f"[Kubeflow] Checkpoint upload failed for {checkpoint_name}: {e}. "
+                            "This may be caused by network issues, S3 outage, or "
+                            "permission changes. Verify S3 connectivity and retry the training job."
+                        ) from e
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)